src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
+        return returnNode.data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> updateNode = findNodeByIndex(index);
+        T oldNode = updateNode.data;
+        updateNode.data = value;
+        return oldNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validIndex(index);
+        Node<T> current = null;
+        T nodeToRemove = null;
+        if (size == 1) {
+            nodeToRemove = head.data;
+            head = null;
+            tail = null;
+            size--;
+            return nodeToRemove;
+        } else if (index == 0) {
+            nodeToRemove = head.data;
+            head = head.next;
+            head.prev = null;
+            size--;
+            return nodeToRemove;
+        } else if (index == size - 1) {
+            nodeToRemove = tail.data;
+            tail = tail.prev;
+            tail.next = null;
+            size--;
+            return nodeToRemove;
+        } else {
+            current = findNodeByIndex(index);
+            nodeToRemove = current.data;
+            unlink(current);
+        }
+        return nodeToRemove;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> nodeToRemove = head;
+        while (nodeToRemove != null) {
+            if (object == nodeToRemove.data
+                    || object != null && object.equals(nodeToRemove.data)) {
+                if (nodeToRemove == head) {
+                    head = head.next;
+                    size--;
+                    return true;
+                }
+                if (nodeToRemove == tail) {
+                    tail = nodeToRemove.prev;
+                    if (tail != null) {
+                        tail.next = null;
+                        size--;
+                        return true;
+                    }
+                } else {
+                    unlink(nodeToRemove);
+                    return true;
+                }
+            }
+            nodeToRemove = nodeToRemove.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        validIndex(index);
+        Node<T> current = null;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        }
+        if (index >= size / 2) {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private void unlink(Node<T> nodeToDelete) {
+        nodeToDelete.prev.next = nodeToDelete.next;
+        nodeToDelete.next.prev = nodeToDelete.prev;
+        size--;
+    }
```

Comment:
```suggestion
    private T unlink(Node<T> x) {
        final T element = x.data;
        Node<T> prev = x.prev;
        Node<T> next = x.next;

        if (prev == null) {
            head = next;
        } else {
            prev.next = next;
            x.prev = null;
        }

        if (next == null) {
            tail = prev;
        } else {
            next.prev = prev;
            x.next = null;
        }

        x.data = null;
        size--;
        return element;
    }
```
let's move your logic to unlink and simplify remove methods